File: apps/api/engine/sarif.py
"""
SARIF 2.1.0 output formatter.
Converts a ScanResult into a SARIF document for GitHub Security tab integration.
"""
from .models import ScanResult, Finding, Severity

_SEVERITY_TO_SARIF_LEVEL: dict[Severity, str] = {
    Severity.CRITICAL: "error",
    Severity.HIGH: "error",
    Severity.MEDIUM: "warning",
    Severity.LOW: "note",
    Severity.INFO: "none",
}

_SEVERITY_TO_SECURITY_SEVERITY: dict[Severity, str] = {
    Severity.CRITICAL: "9.8",
    Severity.HIGH: "7.5",
    Severity.MEDIUM: "5.0",
    Severity.LOW: "2.5",
    Severity.INFO: "0.0",
}


def to_sarif(result: ScanResult, config_path: str = "mcp.json") -> dict:
    rules: list[dict] = []
    rule_ids_seen: set[str] = set()

    for finding in result.findings:
        if finding.check_id not in rule_ids_seen:
            rule_ids_seen.add(finding.check_id)
            rules.append({
                "id": finding.check_id,
                "name": _check_id_to_name(finding.check_id),
                "shortDescription": {"text": finding.title},
                "fullDescription": {"text": finding.detail},
                "helpUri": f"https://mcpaudit.app/checks/{finding.check_id}",
                "properties": {
                    "tags": [finding.owasp.value, "security", "mcp"]
                        + ([finding.cwe_id] if finding.cwe_id else []),
                    "security-severity": _SEVERITY_TO_SECURITY_SEVERITY[finding.severity],
                    "precision": "high",
                    "problem.severity": finding.severity.value,
                },
                "defaultConfiguration": {
                    "level": _SEVERITY_TO_SARIF_LEVEL[finding.severity],
                },
            })

    results: list[dict] = []
    for i, finding in enumerate(result.findings):
        results.append({
            "ruleId": finding.check_id,
            "level": _SEVERITY_TO_SARIF_LEVEL[finding.severity],
            "message": {
                "text": f"{finding.title}\n\n{finding.detail}\n\nRemediation: {finding.remediation}"
            },
            "locations": [
                {
                    "physicalLocation": {
                        "artifactLocation": {
                            "uri": config_path,
                            "uriBaseId": "%SRCROOT%",
                        },
                        "region": {
                            "startLine": 1,
                            "startColumn": 1,
                        },
                    },
                    "logicalLocations": [
                        {
                            "name": finding.server_name,
                            "kind": "member",
                            "fullyQualifiedName": f"mcpServers.{finding.server_name}",
                        }
                    ],
                }
            ],
            "properties": {
                "owasp": finding.owasp.value,
                "server": finding.server_name,
                "engine": finding.engine,
                **({"cwe": finding.cwe_id} if finding.cwe_id else {}),
                **({"attackTactic": finding.attack_tactic} if finding.attack_tactic else {}),
            },
            "fingerprints": {
                "mcpAudit/v1": f"{result.config_hash}/{finding.check_id}/{finding.server_name}/{i}",
            },
        })

    return {
        "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
        "version": "2.1.0",
        "runs": [
            {
                "tool": {
                    "driver": {
                        "name": "MCPAudit",
                        "version": "0.1.0",
                        "informationUri": "https://mcpaudit.app",
                        "rules": rules,
                    }
                },
                "results": results,
                "automationDetails": {
                    "id": result.scan_id,
                    "description": {
                        "text": f"MCPAudit scan {result.scan_id} — {len(result.findings)} findings across {result.summary.servers_scanned} server(s)"
                    },
                },
                "versionControlProvenance": [],
            }
        ],
    }


def _check_id_to_name(check_id: str) -> str:
    prefixes = {
        "SEC": "SecretExposure",
        "SC": "SupplyChain",
        "PI": "PromptInjection",
        "DX": "DataExfiltration",
        "PE": "PrivilegeEscalation",
        "SH": "ShadowServer",
        "AT": "AuditTelemetry",
        "EX": "CodeExecution",
        "LF": "LifecycleScript",
        "CL": "ConfigLevel",
        "EC": "EnvironmentConfig",
    }
    prefix = check_id.split("-")[0]
    suffix = check_id.split("-")[1] if "-" in check_id else ""
    return prefixes.get(prefix, "SecurityCheck") + suffix

File: packages/cli/internal/engine/sarif.go
package engine

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/abhijitdalal26/MCP-Audit/cli/internal/engine/models"
)

var severityToSARIFLevel = map[models.Severity]string{
	models.SeverityCritical: "error",
	models.SeverityHigh:     "error",
	models.SeverityMedium:   "warning",
	models.SeverityLow:      "note",
	models.SeverityInfo:     "none",
}

var severityToSecuritySeverity = map[models.Severity]string{
	models.SeverityCritical: "9.8",
	models.SeverityHigh:     "7.5",
	models.SeverityMedium:   "5.0",
	models.SeverityLow:      "2.5",
	models.SeverityInfo:     "0.0",
}

func checkIDToName(checkID string) string {
	prefixes := map[string]string{
		"SEC": "SecretExposure", "SC": "SupplyChain", "PI": "PromptInjection",
		"DX": "DataExfiltration", "PE": "PrivilegeEscalation", "SH": "ShadowServer",
		"AT": "AuditTelemetry", "EX": "CodeExecution", "LF": "LifecycleScript",
		"CL": "ConfigLevel", "EC": "EnvironmentConfig", "CHAIN": "CrossServerChain",
	}
	parts := strings.SplitN(checkID, "-", 2)
	suffix := ""
	if len(parts) == 2 {
		suffix = parts[1]
	}
	if name, ok := prefixes[parts[0]]; ok {
		return name + suffix
	}
	return "SecurityCheck" + suffix
}

func marshalSARIF(result *models.ScanResult) ([]byte, error) {
	type sarifText struct {
		Text string `json:"text"`
	}
	type sarifArtifact struct {
		URI       string `json:"uri"`
		URIBaseID string `json:"uriBaseId"`
	}
	type sarifRegion struct {
		StartLine   int `json:"startLine"`
		StartColumn int `json:"startColumn"`
	}
	type sarifPhysical struct {
		ArtifactLocation sarifArtifact `json:"artifactLocation"`
		Region           sarifRegion   `json:"region"`
	}
	type sarifLogical struct {
		Name              string `json:"name"`
		Kind              string `json:"kind"`
		FullyQualifiedName string `json:"fullyQualifiedName"`
	}
	type sarifLocation struct {
		PhysicalLocation sarifPhysical  `json:"physicalLocation"`
		LogicalLocations []sarifLogical `json:"logicalLocations"`
	}
	type sarifRuleProps struct {
		Tags             []string `json:"tags"`
		SecuritySeverity string   `json:"security-severity"`
		Precision        string   `json:"precision"`
		ProblemSeverity  string   `json:"problem.severity"`
	}
	type sarifDefaultConfig struct {
		Level string `json:"level"`
	}
	type sarifRule struct {
		ID                   string             `json:"id"`
		Name                 string             `json:"name"`
		ShortDescription     sarifText          `json:"shortDescription"`
		FullDescription      sarifText          `json:"fullDescription"`
		HelpURI              string             `json:"helpUri"`
		Properties           sarifRuleProps     `json:"properties"`
		DefaultConfiguration sarifDefaultConfig `json:"defaultConfiguration"`
	}
	type sarifResultProps struct {
		OWASP        string `json:"owasp"`
		Server       string `json:"server"`
		Engine       string `json:"engine"`
		CWE          string `json:"cwe,omitempty"`
		AttackTactic string `json:"attackTactic,omitempty"`
	}
	type sarifFingerprints struct {
		MCPAuditV1 string `json:"mcpAudit/v1"`
	}
	type sarifResult struct {
		RuleID       string            `json:"ruleId"`
		Level        string            `json:"level"`
		Message      sarifText         `json:"message"`
		Locations    []sarifLocation   `json:"locations"`
		Properties   sarifResultProps  `json:"properties"`
		Fingerprints sarifFingerprints `json:"fingerprints"`
	}

	rules := []sarifRule{}
	seenRules := map[string]bool{}
	results := []sarifResult{}

	for i, f := range result.Findings {
		if !seenRules[f.CheckID] {
			seenRules[f.CheckID] = true
			tags := []string{string(f.OWASP), "security", "mcp"}
			if f.CWEID != "" {
				tags = append(tags, f.CWEID)
			}
			rules = append(rules, sarifRule{
				ID:               f.CheckID,
				Name:             checkIDToName(f.CheckID),
				ShortDescription: sarifText{f.Title},
				FullDescription:  sarifText{f.Detail},
				HelpURI:          fmt.Sprintf("https://mcpaudit.app/checks/%s", f.CheckID),
				Properties: sarifRuleProps{
					Tags:             tags,
					SecuritySeverity: severityToSecuritySeverity[f.Severity],
					Precision:        "high",
					ProblemSeverity:  string(f.Severity),
				},
				DefaultConfiguration: sarifDefaultConfig{Level: severityToSARIFLevel[f.Severity]},
			})
		}

		msg := fmt.Sprintf("%s\n\n%s\n\nRemediation: %s", f.Title, f.Detail, f.Remediation)
		props := sarifResultProps{
			OWASP:        string(f.OWASP),
			Server:       f.ServerName,
			Engine:       f.Engine,
			CWE:          f.CWEID,
			AttackTactic: f.AttackTactic,
		}
		results = append(results, sarifResult{
			RuleID:  f.CheckID,
			Level:   severityToSARIFLevel[f.Severity],
			Message: sarifText{msg},
			Locations: []sarifLocation{{
				PhysicalLocation: sarifPhysical{
					ArtifactLocation: sarifArtifact{URI: "mcp.json", URIBaseID: "%SRCROOT%"},
					Region:           sarifRegion{StartLine: 1, StartColumn: 1},
				},
				LogicalLocations: []sarifLogical{{
					Name:               f.ServerName,
					Kind:               "member",
					FullyQualifiedName: fmt.Sprintf("mcpServers.%s", f.ServerName),
				}},
			}},
			Properties: props,
			Fingerprints: sarifFingerprints{
				MCPAuditV1: fmt.Sprintf("%s/%s/%s/%d", result.ConfigHash, f.CheckID, f.ServerName, i),
			},
		})
	}

	doc := map[string]any{
		"$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
		"version": "2.1.0",
		"runs": []map[string]any{{
			"tool": map[string]any{
				"driver": map[string]any{
					"name":           "MCPAudit",
					"version":        "0.1.0",
					"informationUri": "https://mcpaudit.app",
					"rules":          rules,
				},
			},
			"results": results,
			"automationDetails": map[string]any{
				"id": result.ScanID,
				"description": map[string]any{
					"text": fmt.Sprintf("MCPAudit scan %s — %d findings across %d server(s)",
						result.ScanID, len(result.Findings), result.Summary.ServersScanned),
				},
			},
			"versionControlProvenance": []any{},
		}},
	}

	return json.MarshalIndent(doc, "", "  ")
}
